compiler/ir/backend.common/src/org/jetbrains/kotlin/ir/util/IrTypeUtils.kt

```diff
@@ -119,3 +119,28 @@ fun IrType.substitute(substitutionMap: Map<IrTypeParameterSymbol, IrType>): IrTy
         newAnnotations
     )
 }
+
+fun getImmediateSupertypes(type: IrSimpleType): List<IrSimpleType> {
+    val originalSuperTypes = type.classOrNull?.superTypes() ?: emptyList()
+    val args = type.arguments.mapNotNull { (it as? IrTypeProjection)?.type }
+    return originalSuperTypes
+        .filter { it.classOrNull != null }
+        .map { superType ->
+            superType.substitute(superType.classOrNull!!.owner.typeParameters, args) as IrSimpleType
+        }
+}
+
+fun collectAllSupertypes(type: IrSimpleType, result: HashSet<IrSimpleType>) {
+    val immediateSuperTypes = getImmediateSupertypes(type)
+    result.addAll(immediateSuperTypes)
+    immediateSuperTypes.forEach {
+        collectAllSupertypes(it, result)
+    }
+}
+
+fun getAllSuperTypes(type: IrSimpleType): HashSet<IrSimpleType> {
```

Comment:
`HashSet` -> `MutableSet`

Reply:
Also "supertype" is one word, so please rename to `getAllSupertypes`

Reply:
Also, a type cannot have supertypes, but a class can. Let's change the parameter of these methods to `IrClass`, and either use `defaultType` or `typeParameters` in the implementation, to obtain the type parameters of the target class